core/src/test/scala/org/apache/spark/ui/memory/MemoryListenerSuite.scala

```diff
@@ -0,0 +1,258 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ui.memory
+
+import org.apache.spark._
+import org.apache.spark.executor._
+import org.apache.spark.scheduler._
+import org.apache.spark.scheduler.cluster._
+
+class MemoryListenerSuite extends SparkFunSuite with LocalSparkContext {
+  test("test HashMap size for MemoryListener") {
+    val listener = new MemoryListener
+    val execId1 = "exec-1"
+    val execId2 = "exec-2"
+
+    (1 to 2).foreach { i =>
+      listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(i))
+      listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(i))
+    }
+    // stages are all completed, no activeStages now
+    assert(listener.activeStagesToMem.isEmpty)
+
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId1, new ExecutorMetrics))
+    // ExecutorMetrics is not related with Stages directly
+    assert(listener.activeStagesToMem.isEmpty)
+
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(3))
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId2, new ExecutorMetrics))
+    // totally 2 executors updated their metrics
+    assert(listener.activeExecutorIdToMem.size === 2)
+    assert(listener.activeStagesToMem.size === 1)
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(3))
+
+    assert(listener.activeStagesToMem.isEmpty)
+    assert(listener.completedStagesToMem.size === 3)
+    assert(listener.activeExecutorIdToMem.size === listener.latestExecIdToExecMetrics.size)
+    assert(listener.removedExecutorIdToMem.isEmpty)
+  }
+
+  test("test first stage with no executor metrics update") {
+    val listener = new MemoryListener
+    val execId1 = "exec-1"
+
+    listener.onExecutorAdded(
+      SparkListenerExecutorAdded(0L, execId1, new ExecutorInfo("host1", 1, Map.empty)))
+
+    // stage 1, no metrics update
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(1))
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(1))
+
+    // stage 2, with one metrics update
+    listener.onStageSubmitted(MemoryListenerSuite.createStageStartEvent(2))
+    val execMetrics = MemoryListenerSuite.createExecutorMetrics("host-1", 0L, 20, 10)
+    listener.onExecutorMetricsUpdate(MemoryListenerSuite.createExecutorMetricsUpdateEvent(
+      execId1, execMetrics))
+    listener.onStageCompleted(MemoryListenerSuite.createStageEndEvent(2))
+
+    val mapForStage1 = listener.completedStagesToMem.get((1, 0)).get
```

Comment:
nit: rather than `hashmap.get(x).get` you can just do `hashmap(x)`, here and elsewhere